asserts/snapasserts/validation_sets.go

```diff
@@ -75,54 +75,91 @@ type revConstraint struct {
 	asserts.ValidationSetSnap
 }
 
-func (c *snapContraints) conflict() (conflicting []string, err error) {
+func (c *snapContraints) conflict() *snapConflictsError {
 	if c.presence != presConflict {
-		return nil, nil
+		return nil
 	}
 
 	const dontCare asserts.Presence = ""
-	whichSets := func(rcs []*revConstraint, presence asserts.Presence) string {
+	whichSets := func(rcs []*revConstraint, presence asserts.Presence) []string {
 		which := make([]string, 0, len(rcs))
 		for _, rc := range rcs {
 			if presence != dontCare && rc.Presence != presence {
 				continue
 			}
-			conflicting = append(conflicting, rc.validationSetKey)
 			which = append(which, rc.validationSetKey)
 		}
 		if len(which) == 0 {
-			return ""
+			return nil
 		}
 		sort.Strings(which)
+		return which
+	}
+
+	byRev := make(map[snap.Revision][]string, len(c.revisions))
+	for r := range c.revisions {
+		pres := dontCare
+		switch r {
+		case invalidPresRevision:
+			pres = asserts.PresenceInvalid
+		case unspecifiedRevision:
+			pres = asserts.PresenceRequired
+		}
+		which := whichSets(c.revisions[r], pres)
+		if len(which) != 0 {
+			byRev[r] = which
+		}
+	}
+
+	return &snapConflictsError{
+		name:      c.name,
+		revisions: byRev,
+	}
+}
+
+type snapConflictsError struct {
+	name string
+	// revisions maps revisions to validation-set keys of the sets
+	// that are in conflict over the revision.
+	// * unspecifiedRevision is used for validation-sets conflicting
+	//   on the snap by requiring it but without a revision
+	// * invalidPresRevision is used for validation-sets that mark
+	//   presence as invalid
+	// see snapContraints.revisions as well
+	revisions map[snap.Revision][]string
+}
+
+func (e *snapConflictsError) Error() string {
+	whichSets := func(which []string) string {
 		return fmt.Sprintf("(%s)", strings.Join(which, ","))
 	}
 
-	msg := fmt.Sprintf("cannot constraint snap %q", c.name)
+	msg := fmt.Sprintf("cannot constraint snap %q", e.name)
```

Comment:
Just noticed this now, should this be "cannot constrain ..." (no "t")?

Reply:
oops, thanks